logging: log_output: syslog timestamp

maass-hamburg opened the pull request:
- add test for log output of net backend (syslog)
- add missing dependency to log backend net
- use `gmtime_r()` for the timestamp in log_output, even if `strftime()` is not available 
- add testcase for that

maass-hamburg committed:
tests: logging: log_output: net: add test for syslog output

add test for syslog log output.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
logging: backend: net: add missing dependency

add missing dependency of NET_SOCKETS to LOG_BACKEND_NET.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

nordic-krch commented on tests/subsys/logging/log_output_net/src/log_output_test.c:
```diff
@@ -0,0 +1,79 @@
+			   0, flags);
+
+	mock_buffer[mock_len] = '\0';
+	printk("%s", exp_str);
```
those `printk` entries can be removed. Especially that they are expected to be equal. Maybe `zassert_str_equal` can have print of those strings in case of failure?

cfriedt commented on subsys/logging/log_output.c:
```diff
@@ -158,6 +155,17 @@ static inline bool is_leap_year(uint32_t year)
 static void __attribute__((unused)) get_YMD_from_seconds(uint64_t seconds,
 			struct YMD_date *output_date)
 {
+#if defined(CONFIG_COMMON_LIBC_GMTIME_R) || defined(CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R)
```
```suggestion
#if defined(CONFIG_COMMON_LIBC_GMTIME_R) || defined(CONFIG_POSIX_C_LANG_SUPPORT_R)
```

maass-hamburg replied:
no, we need to use `CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R`, because that is true, if the toolchain provides `gmtime_r`. `CONFIG_POSIX_C_LANG_SUPPORT_R` is just there to enable `CONFIG_COMMON_LIBC_GMTIME_R`( which will provide `gmtime_r`),  if `CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R` is not true.

cfriedt replied:
> no, we need to use `CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R`, because that is true, if the toolchain provides `gmtime_r`.

Half correct.

`CONFIG_POSIX_C_LANG_SUPPORT_R` is just there to enable `CONFIG_COMMON_LIBC_GMTIME_R`( which will provide `gmtime_r`), if `CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R` is not true.

`CONFIG_TC_PROVIDES_*` options are exclusively for negotiation between the toolchain / libc and the posix api. 

The application should still select `CONFIG_POSIX_C_LANG_SUPPORT_R` which will work the same independent of the libc.

So the application in question should select `CONFIG_POSIX_C_LANG_SUPPORT_R=y`.

cfriedt replied:
Technically, you can also use `_POSIX_C_LANG_SUPPORT_R` here as well @maas-hamburg

cfriedt commented on subsys/logging/log_output.c:
```diff
@@ -158,6 +155,17 @@ static inline bool is_leap_year(uint32_t year)
+	output_date->year = tm_timestamp.tm_year + 1900;
+	output_date->month = tm_timestamp.tm_mon + 1;
+	output_date->day = tm_timestamp.tm_mday;
+#else  /* CONFIG_COMMON_LIBC_GMTIME_R || CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R */
```
```suggestion
#else  /* CONFIG_COMMON_LIBC_GMTIME_R || CONFIG_POSIX_C_LANG_SUPPORT_R */
```

cfriedt commented on subsys/logging/log_output.c:
```diff
@@ -188,6 +196,7 @@ static void __attribute__((unused)) get_YMD_from_seconds(uint64_t seconds,
 	}
 
 	output_date->day += seconds / SECONDS_IN_DAY;
+#endif /* CONFIG_COMMON_LIBC_GMTIME_R || CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R*/
```
```suggestion
#endif /* CONFIG_COMMON_LIBC_GMTIME_R || CONFIG_POSIX_C_LANG_SUPPORT_R*/
```

maass-hamburg committed:
lib: posix: set default of POSIX_C_LANG_SUPPORT_R

enable POSIX_C_LANG_SUPPORT_R by default if
the functions are already provided by the toolchain.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
lib: libc: select POSIX_C_LANG_SUPPORT_R

also select POSIX_C_LANG_SUPPORT_R, so we can check
that, to see if gmtime_r() is available.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
logging: remove part for when gmtime_r is not available

now that CONFIG_POSIX_C_LANG_SUPPORT_R will
add a custom implementation for gmtime_r() if that
is not provided by the toolchain, we can simply depend on
that, instead of using out own.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
I noticed, that we only the two cases where we checked for `CONFIG_REQUIRES_FULL_LIBC`, because of `gmtime_r` and not because of `strftime`. `strftime` is part of the ISO C standard. Now we depend on `CONFIG_POSIX_C_LANG_SUPPORT_R` for syslog,  `CONFIG_LOG_OUTPUT_FORMAT_DATE_TIMESTAMP` and `CONFIG_LOG_OUTPUT_FORMAT_ISO8601_TIMESTAMP`, that means we don't need other case where we don't have `gmtime_r` and instead use the common implementation.

maass-hamburg committed:
lib: libc: newlib: Don't replace newlib libc time _r functions

Select TC_PROVIDES_POSIX_C_LANG_SUPPORT_R to keep Zephyr from
including the common libc implementation of the various _r APIs.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
logging: log_output: move partly unused code

move code, that is only needed for one case
to that case.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

kartben commented:
This PR is causing a test failure in main

```
west twister -p qemu_cortex_m0/nrf51822 -s logging.backend.uart.single --no-detailed-test-id
```

git bisect points to commit 5d2068f4919acc90342438c883eaf3a8ccc79f18 as the culprit.

maass-hamburg commented:
> This PR is causing a test failure in main
> 
> ```
> west twister -p qemu_cortex_m0/nrf51822 -s logging.backend.uart.single --no-detailed-test-id
> ```
> 
> git bisect points to commit [5d2068f](https://github.com/zephyrproject-rtos/zephyr/commit/5d2068f4919acc90342438c883eaf3a8ccc79f18) as the culprit.

@kartben I gone back in main by rebaising and removed the commits from this PR and that Test was still failing.

maass-hamburg commented:
@kartben I now even tested it with the branch of this PR. The test was failing after the commits of this PR. And was also failing on https://github.com/zephyrproject-rtos/zephyr/commit/302683ef7f21284c3ea90b8a24723ff56d536823, which was the last commit from main in this branch before my commits
